fuel-core/src/database/vm_database.rs

```diff
@@ -142,28 +158,1179 @@ impl InterpreterStorage for VmDatabase {
+                    for _ in range_count..range {
+                        results.push(None);
+                        current_key =
+                            current_key.checked_add(1.into()).ok_or_else(|| {
+                                Error::Other(anyhow!(
+                                    "range op exceeded available keyspace"
+                                ))
+                            })?;
+                    }
+                    // Iterator no longer useful, return
+                    return Ok(results)
+                } else if state_key != current_key {
+                    while (state_key != current_key) && (range_count < range) {
+                        // Iterator moved beyond next expected key, push none and increment range
+                        // count until we find the current key
+                        results.push(None);
+                        range_count += 1;
+                        current_key =
+                            current_key.checked_add(1.into()).ok_or_else(|| {
+                                Error::Other(anyhow!(
+                                    "range op exceeded available keyspace"
+                                ))
+                            })?;
+                    }
+                }
+                // State key matches, put value into results
+                if state_key == current_key {
+                    results.push(Some(Cow::Owned(value)));
+                    current_key = current_key.checked_add(1.into()).ok_or_else(|| {
+                        Error::Other(anyhow!("range op exceeded available keyspace"))
+                    })?;
+                }
+            } else {
+                // No iterator returned, populate with None until end of range
+                let range_to_fill = range_count..range;
+                for _ in range_to_fill {
+                    results.push(None);
+                    range_count += 1;
+                    current_key = current_key.checked_add(1.into()).ok_or_else(|| {
+                        Error::Other(anyhow!("range op exceeded available keyspace"))
+                    })?;
+                }
+            };
+            range_count += 1;
+        }
+
+        Ok(results)
     }
 
     fn merkle_contract_state_insert_range(
         &mut self,
-        _contract: &ContractId,
-        _start_key: &Bytes32,
-        _values: &[Bytes32],
+        contract_id: &ContractId,
+        start_key: &Bytes32,
+        values: &[Bytes32],
     ) -> Result<Option<()>, Self::DataError> {
-        unimplemented!()
+        let mut found_unset = false;
+
+        let mut current_key = U256::from_big_endian(start_key.as_ref());
+        let mut key_bytes = [0u8; 32];
+
+        let transaction = self.database.transaction();
+        let transaction_db = transaction.deref();
+
+        for value in values {
+            current_key.to_big_endian(&mut key_bytes);
```

Comment:
Is this actually correct? Won't rocks db be storing the `Bytes32` keys ordered by the actual bytes in the array not as a number?

Reply:
The compiler will decide which endianess the keys are in when incrementing slot keys, and they confirmed that they plan to use big endian


cc @mohammadfawaz